### DISCUSSION TITLE
Commander X16 Banking (compiling large project)

### DISCUSSION THREAD

**Discussion by voidstar78** [cc65/cc65#2164] [General]
I have some existing straightforward C code.  It has all compiled.  There are some warnings that might still need to be addressed. For example:
mt.c:125: Warning: Shift count '30' >= width of type

But ignoring that for now..... I also had to turn off optimization because there was a branch distance issue during the ca65 assembly. Specifically:
game.s:6889: Error: Range error (131 not in [-128..127])

I'm trying to compile to the cx16 target.  So I swapped in the cx16-bank.cfg (actually didn't know how to redirect to that .cfg so I just copied the cx16-bank.cfg over into the cx16.cfg).  Doing like this:
ld65 --target cx16 --obj compress.o game.o mt.o parser.o tables.o villains.o _data.o --lib ..\lib\cx16.lib -o main_cx16.out

Which ends up reporting:
ld65: Warning: F:\X16\cc65-snapshot-win32\cfg/cx16.cfg:17: Segment 'CODE' overflows memory area 'MAIN' by 66614 bytes
ld65: Error: Cannot generate most of the files due to memory area overflow

The code, assembly, and object files come out to these sizes.
3,795	compress.c
193,282	game.c
4,660	mt.c
71,638	parser.c
94,094	tables.c
41,836	villains.c
51,234	_data.c
460,539	  total bytes
	
17,185	compress.s
653,112	game.s
16,778	mt.s
259,048	parser.s
94,337	tables.s
154,038	villains.s
59,340	_data.s
1,253,838   total bytes
	
18,959	compress.o
748,002	game.o
24,014	mt.o
307,317	parser.o
106,449	tables.o
184,550	villains.o
81,900	_data.o
1,471,191   total bytes

So.... is the 1.4mb of object code itself too large?  Banking is new territory for me.  And I've never used the HIMEM stuff in cc65.  It looks to me like cx16-bank.cfg has 15 banks defined.  For example:

    BRAM0A:     file = "%O.0A",           start = __BANKRAMSTART__,     size = __BANKRAMSIZE__, bank = $0A;
    BRAM0BADDR: file = "%O.0B",           start = __BANKRAMSTART__ - 2, size = $0002;

Where the start and size seem right to me:

    __BANKRAMSTART__: type = export, value = $A000;
    __BANKRAMSIZE__:  type = weak,   value = $2000; # 8K banked RAM

So thoughts on what to try next?

**Comment by mrdudz** [comment]
Banking does not happen automatically with cc65. So the largest project you can compile "as is" cant be larger than 64k (minus whatever workspace the program needs)

**Reply by voidstar78** [reply]
Makes sense, it would be some kind of "OS" that has to provide interfaces for malloc to know where any available free space is located at.  And then if you loop across that entire malloc content, the compiler can't know which portions are in what bank.  

Mostly I was curious why the error is saying 66K bytes needed - by the size of the .o files, it seems a lot more code space would be needed?

**Reply by wizardmanannan** [reply]
Hi VoidStar,
I may be a pioneer in this area.

For the past year I have been porting the Meka Sierra AGI interpreter to CX16.

If you don't already know; AGI is the script which supports old adventure games like King's Quest.

I was lucky enough to find Meka by Lance Ewing, which was written in C89.

Like I have mentioned recently in the forum I have a interpreter that works well enough to run the intro scene in King III (no text or sprites yet).

https://www.youtube.com/watch?v=upsdRqa_NhY

I think that my code may particularly help you with Banking. As the vast majority of my code in on the banks.

See my repo at https://github.com/wizardmanannan/CX16Agi

Here is a simple example of memory banking I provided to a documentation writer: https://www.dropbox.com/scl/fi/7rwybzo69q5dsjqg9eb86/exampleBankCall-1.c?rlkey=jkrln3p5uc0ooqvdhi68l05di&dl=0

I also changed the config file to allow for extra banks to be used for code: https://github.com/wizardmanannan/CX16Agi/blob/main/cx16-bank.cfg.

My suggestion it to compare it to the standard file with a compare tools to see the differences. 

I also suggest you look at my bankedRamInit function in https://github.com/wizardmanannan/CX16Agi/blob/main/src/memoryManager.c

I found that sometimes you need to skip the first two bytes in the bank ram data file, and sometimes you don't as you can see from my logic.

**Reply by rofl0r** [reply]
@wizardmanannan: this platform (github) provides its own pastebin service, gist.github.com, and allows file attachment in comments so you using dropbox for your example code is really annoying, especially for someone finding this in 2 years when your link has expired, dropbox went bankrupt, or they start loading so much javascript that your browser crashes.

**Comment by mrdudz** [comment]
The object files contain much more info than just the assembled code (debug info, linker info etc)

**Comment by wizardmanannan** [comment]
@rofl0r
Hi,
I am new to GitHub, I didn't release there was a pastebin service.

Here is a link to the file in the pastebin:
https://gist.github.com/wizardmanannan/d0f164efa92b61727380905ef268018f
